### PR TITLE
Properly check header before extracting the bearer token

### LINKED ISSUE #42591: StringIndexOutOfBoundsException on OAuth2AuthMechanism
### Describe the bug

I'm seeing a `StringIndexOutOfBoundsException` when the Authorization header is not valid

### Expected behavior

It should return an invalid header

### Actual behavior

throwing `StringIndexOutOfBoundsException`

### How to Reproduce?

We have an application deployed and we get the usual random worldpress requests trying to access our system, most of the requests don't cause any issues but from time to time we get a 500, the last time we saw the 500 error we noticed this error message on our logs `java.lang.StringIndexOutOfBoundsException: begin 7, end 6, length 6`. Looking at the entire error log we noticed that the error was being thrown inside `elytron-security-oauth2` more specifically in [here](https://github.com/quarkusio/quarkus/blob/main/extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/auth/OAuth2AuthMechanism.java#L45).

What is causing this error is the fact that the _"malicious"_ user is not setting the Bearer token correctly. When the user sends an empty `Bearer ` the code throws an error. 

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

21

### Quarkus version or git rev

3.14.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

The error is caused by the string size not being checked before the substring method is called.

Potential solution

```
String authHeader = context.request().headers().get("Authorization");
String bearerToken = null;
if (authHeader != null && authHeader.length() > 7) {  // Ensure authHeader is long enough
    bearerToken = authHeader.substring(7);
}
if (bearerToken != null && !bearerToken.isEmpty()) {
    // Install the OAuth2 principal as the caller
    return identityProviderManager
            .authenticate(new TokenAuthenticationRequest(new TokenCredential(bearerToken, "bearer")));
}
// No suitable header has been found in this request,
return Uni.createFrom().nullItem();
```

### PR DESCRIPTION
**Pull request by gsmet**
Fixes #42591

### CODE REVIEW

**Review by sberyozkin — APPROVED** [review]
Thanks @gsmet

File: extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/auth/OAuth2AuthMechanism.java [hunk 1]
```diff
@@ -42,15 +44,17 @@ public class OAuth2AuthMechanism implements HttpAuthenticationMechanism {
     public Uni<SecurityIdentity> authenticate(RoutingContext context,
             IdentityProviderManager identityProviderManager) {
         String authHeader = context.request().headers().get("Authorization");
-        String bearerToken = authHeader != null ? authHeader.substring(7) : null;
-        if (bearerToken != null) {
-            // Install the OAuth2 principal as the caller
-            return identityProviderManager
-                    .authenticate(new TokenAuthenticationRequest(new TokenCredential(bearerToken, "bearer")));
 
+        if (authHeader == null || !authHeader.startsWith(BEARER_PREFIX)) {
```
Comment (sberyozkin) [inline extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/auth/OAuth2AuthMechanism.java:48]: May be tests have some spaces around... It all looks fine

Comment (gsmet) [reply extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/auth/OAuth2AuthMechanism.java:48]: The test I checked is using `Bearer: <token>` with a colon which is AFAICS incorrect. I'll go fix it I suppose.

Comment (gsmet) [reply extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/auth/OAuth2AuthMechanism.java:48]: I added an additional commit, I'm in the middle of something and can't test my tree.

**Comment by gsmet** [comment]
@sberyozkin I let you check that I’m correct regarding the tests. All the examples I could find didn’t have a colon.

**Review by michalvavrik — APPROVED** [review]

**Comment by sberyozkin** [comment]
Thanks @gsmet This extension was added by Loic long time ago, he did not like the idea of depending on smallrye-jwt. This extension is light weight indeed but it only supports the remote token introspection via Elytron AFAIK

**Comment by sberyozkin** [comment]
Thanks @michalvavrik

**Comment by sberyozkin** [comment]
@gsmet @michalvavrik, FYI, https://quarkus.io/guides/security-authentication-mechanisms#oidc-jwt-oauth2-comparison

**Pull request merged**